### PR TITLE
Promoted properties support

### PR DESCRIPTION
**Pull request by kukulich**

### CODE REVIEW

File: src/Reflection/ReflectionParameter.php [hunk 1]
```diff
@@ -449,6 +449,11 @@ public function canBePassedByValue(): bool
         return ! $this->isPassedByReference();
     }
 
+    public function isPromoted(): bool
+    {
+        return $this->node->flags !== 0;
```
Comment (Ocramius) [inline src/Reflection/ReflectionParameter.php:454]: Is there a more precise flag value to check? :thinking:

**Review by Ocramius — APPROVED** [review]
This turned out to be much cleaner than I though :O

File: src/Reflection/ReflectionClass.php [hunk 2]
```diff
@@ -664,21 +665,52 @@ public function getImmediateProperties(?int $filter = null): array
...
+                        $this->declaringNamespace,
+                        $this,
+                        $this,
+                        false,
+                    );
+                    $properties[$property->getName()] = $property;
+                }
+            }
+
+            foreach ($this->node->getMethods() as $methodNode) {
```
Comment (Ocramius) [inline src/Reflection/ReflectionClass.php:683]: Is there a faster way to find the ctor, than looping through these? (fine also if not)

Comment (kukulich) [reply src/Reflection/ReflectionClass.php:683]: I think this is the most optimized way. We don't need to create any reflection, just iterate existing AST.

File: src/Reflection/ReflectionParameter.php (hunk 1 above)
Comment (kukulich) [reply src/Reflection/ReflectionParameter.php:454]: Unfortunately not: https://github.com/nikic/PHP-Parser/blob/master/lib/PhpParser/Node/Param.php#L40

Comment (Ocramius) [reply src/Reflection/ReflectionParameter.php:454]: Yeah, only found one cryptic usage: `$this->semStack[$stackPos-(6-2)]` :D 

Let's roll it as-is :+1:

**Comment by Ocramius** [comment]
Awesome++ 

Thanks @kukulich!

**Pull request merged**